src/kernel/arch/aarch64/arch.zig

```diff
@@ -35,50 +36,38 @@ pub fn initTTY(boot_payload: BootPayload) TTY {
-    mmio.write(mmio_addr, mmio.Register.GPIO_PULL_CLK, (1 << 14) | (1 << 15));
-    // Apply pull up/down change for pins 14 and 15
-    mmio.write(mmio_addr, mmio.Register.GPIO_PULL_CLK, 0);
-    // Clear pending uart interrupts
-    mmio.write(mmio_addr, mmio.Register.UART_INT_CONTROL, 0x7FF);
-
-    // For models after rpi 2 (those supporting aarch64) the uart clock is dependent on the system clock
-    // so set it to a known constant rather than calculating it
-    const cmd: []const volatile u32 align(16) = &[_]u32{ 36, 0, 0x38002, 12, 8, 2, SYSTEM_CLOCK, 0, 0 };
-    const cmd_addr = (@intCast(u32, @ptrToInt(&cmd)) & ~@as(u32, 0xF)) | 8;
-    // Wait until the mailbox isn't busy
-    while ((mmio.read(mmio_addr, mmio.Register.MBOX_STATUS) & 0x80000000) != 0) {}
-    mmio.write(mmio_addr, mmio.Register.MBOX_WRITE, cmd_addr);
-    // Wait for the correct response
-    while ((mmio.read(mmio_addr, mmio.Register.MBOX_STATUS) & 0x40000000) != 0 or mmio.read(mmio_addr, mmio.Register.MBOX_READ) != cmd_addr) {}
-
-    // Setup baudrate
-    const divisor: f32 = @intToFloat(f32, SYSTEM_CLOCK) / @intToFloat(f32, 16 * UART_BAUD_RATE);
-    const divisor_int: u32 = @floatToInt(u32, divisor);
-    const divisor_frac: u32 = @floatToInt(u32, (divisor - @intToFloat(f32, divisor_int)) * 64.0 + 0.5);
-    mmio.write(mmio_addr, mmio.Register.UART_BAUD_INT, divisor_int);
-    mmio.write(mmio_addr, mmio.Register.UART_BAUD_FRAC, divisor_frac);
-
-    // Enable FIFO, 8 bit words, 1 stop bit and no parity bit
-    mmio.write(mmio_addr, mmio.Register.UART_LINE_CONTROL, 1 << 4 | 1 << 5 | 1 << 6);
-    // Mask all interrupts
-    mmio.write(mmio_addr, mmio.Register.UART_INT_MASK, 1 << 1 | 1 << 4 | 1 << 5 | 1 << 6 | 1 << 7 | 1 << 8 | 1 << 9 | 1 << 10);
-    // Enable UART0 receive and transmit
-    mmio.write(mmio_addr, mmio.Register.UART_CONTROL, 1 << 0 | 1 << 8 | 1 << 9);
-
+    is_qemu = interrupts.mrs("cntfrq_el0") != 0;
+    if (!is_qemu) {
+        pinSetFunction(14, .AlternateFunction5);
+        pinSetPull(14, .None);
+        pinSetFunction(15, .AlternateFunction5);
+        pinSetPull(15, .None);
+        mmio.write(mmio_addr, .AUX_ENABLES, 1);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_LCR_REG, 3);
+        mmio.write(mmio_addr, .AUX_MU_MCR_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IIR_REG, 0xc6);
+        mmio.write(mmio_addr, .AUX_MU_BAUD_REG, 270);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 3);
+    }
     return .{
         .write = uartWriteByte,
     };
 }
 
-fn uartWriteByte(byte: u8) void {
-    // Wait until the UART is ready to transmit
-    while ((mmio.read(mmio_addr, mmio.Register.UART_FLAGS) & (1 << 5)) != 0) {}
-    mmio.write(mmio_addr, mmio.Register.UART_DATA, byte);
+pub fn uartWriteByte(byte: u8) void {
+    if (byte == 10) {
+        uartWriteByte(13);
+    }
```

Comment:
What's the purpose of this check?

Reply:
This adds a carriage return to a line feed. Should I add a comment or is there a better place to do this?

Reply:
Hmm sending a carriage return as well sounds very windows-like. Is this standard in serial communication?

Reply:
Somewhere along the line a carriage return is needed for the terminal to move the print head to the left position. Line feed only advances the paper. This is real teletype stuff.

This is needed for an attached tty or a terminal emulator like minicom.

Reply:
OK that makes sense. You could add a brief comment to explain that.

Reply:
Done: 43da14b643f868eb9a12dc80e4c4ddcb8cd21117